### PR TITLE
Add a Bessemer VSCode Remote script.

### PR DESCRIPTION
**Pull request by jkwmoore**
As stated above.

### CODE REVIEW

File: start_vscode_bessemer.sh [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+# check if VSC_WAITING_INTERVAL is an integer
+if ! [[ "$VSC_WAITING_INTERVAL" =~ ^[0-9]+$ ]]; then
+        echo -e "Error: $VSC_WAITING_INTERVAL -> Waiting time interval [seconds] must be an integer, please try again\n"
+        display_help
+else
+    echo -e "Setting waiting time interval for checking the start of the job to $VSC_WAITING_INTERVAL seconds"
+fi
+
+# set modules
+VSC_MODULE_COMMAND="apps/vscode-server/4.2.0/binary dev/git/2.35.2/gcc-4.9.4"
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:279]: This will need amending to the Bessemer module load.

Comment (jkwmoore) [reply start_vscode_bessemer.sh:279]:
```suggestion
VSC_MODULE_COMMAND="module load vscode-server/4.2.0/binary git/2.28.0-GCCcore-10.2.0-nodocs"
```

File: start_vscode_bessemer.sh [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+$0: Script to start a VSCode remote server on Bessemer from a local computer
+
+Usage: start_vscode_bessemer.sh [options]
+
+Options:
+
+        -u | --username       USERNAME                  TUoS username for SSH connection to Bessemer
+        -W | --runtime        RUN_TIME                  Run time limit for the code-server in hours and minutes HH:MM
+        -n | --numcpus        NUM_CPUS_PER_TASK         Number of CPU cores per task     
+        -m | --memory         MEM_PER_JOB               Memory limit in GB per job. (RAM) Ex. 4 cores *4G = 16
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:79]: This needs amending to Bessemer parlance (memory per node.)

File: start_vscode_bessemer.sh [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+#                                                                             #
+#  Main author    : Samuel Fux                                                #
+#  Contributions  : Andreas Lugmayr                                           #
+#  Date           : October 2021                                              #
+#  Location       : ETH Zurich                                                #
+#  Version        : 0.1                                                       #
+#  Change history :                                                           #
+#                                                                             #
+#  28.10.2021    Initial version of the script based on Jupyter script        #
+#  Forked and edited for TUoS by J.Moore                                      #
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:16]: You may wish to add change history here.

File: start_vscode_bessemer.sh [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+
+        ./start_vscode_bessemer.sh -u te1st -n 4 -W 04:00:00 -m 4
+
+        ./start_vscode_bessemer.sh --username te1st --numcpus 2 --runtime 01:30:00 --memory 2
+
+        ./start_vscode_bessemer.sh -c $HOME/.vsc_config
+
+Format of configuration file:
+
+VSC_USERNAME=""    # TUoS username for SSH connection to Bessemer
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:101]: Alignment needed.

File: start_vscode_bessemer.sh (hunk 2 above)
Comment (jkwmoore) [reply start_vscode_bessemer.sh:79]: Adjusting the `VSC_MEM_PER_JOB` variable to `VSC_MEM_PER_NODE` may also be wise.

File: start_vscode_bessemer.sh [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+
+###############################################################################
+# Check required SSL certs exist                                              #
+###############################################################################
+
+SSLCERT=$(ssh $VSC_SSH_OPT "[ -e ~/.ssl/vscoderemote/vscode_remote_ssl-server-cert.pem ] && echo 1 || echo 0")
+SSLCERTKEY=$(ssh $VSC_SSH_OPT "[ -e ~/.ssl/vscoderemote/private/vscode_remote_ssl-server-key.pem ] && echo 1 || echo 0")
+
+if [[ "$SSLCERT" == 0 ]] || [[ "$SSLCERTKEY" == 0 ]] ; then
+        echo -e "Missing SSL certificate or key. Exiting! Please 'module load apps/vscode-server/4.2.0/binary' run SSL setup step 'setup_ssl_ca_server_client.sh' first! "
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:335]:
```suggestion
        echo -e "Missing SSL certificate or key. Exiting! Please 'module load vscode-server/4.2.0/binary' and run SSL setup step 'setup_ssl_ca_server_client.sh' first! "
```

File: start_vscode_bessemer.sh [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+
+# This is being done this way for reasons.
+ssh -T $VSC_SSH_OPT "sed -i '/^password:/d' ~/.config/code-server/config.yaml && echo 'password: $VSCPASS' > ~/.config/code-server/config.yaml"
+
+# run the code-server job on Bessemer and save the ip of the compute node in the file vscip in the home directory of the user on Bessemer
+echo -e "Connecting to $VSC_HOSTNAME to start the code-server in a batch job"
+# FIXME: save jobid in a variable, that the script can kill the batch job at the end
+echo -e "Connection command:"
+echo -e "==================================================================================="
+echo -e "ssh ${VSC_SSH_OPT} sbatch --export=ALL --cpus-per-task=${VSC_CPUS_PER_TASK} -t=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_JOB}G ${VSC_SNUM_GPU}"
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:354]: We may need to add a warning on both modules / scripts that the default behavior is set to inherit the environment from the users' login session (i.e. .bashrc) even though strictly we are doing this intentionally to ensure their personal software  (if they have any)  / environment is also loaded.

File: start_vscode_bessemer.sh [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+echo -e "===================================================================================\n"
+ssh ${VSC_SSH_OPT} sbatch -J VSCodeServer --cpus-per-task=${VSC_CPUS_PER_TASK} -t=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_JOB}G ${VSC_SNUM_GPU} << ENDSBATCH
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:356]: Line 356 is not matching 354. These need to be clones of one another. Suggest we add the export all into 356 and the name into 354.

File: start_vscode_bessemer.sh [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+source \${HOME}/.bashrc
+module load $VSC_MODULE_COMMAND
+export XDG_RUNTIME_DIR="\$HOME/vsc_runtime"
+VSC_IP_REMOTE="\$(hostname)"
+#VSC_PORT_REMOTE="8899"
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:361]: This line can be removed from both scripts.

File: start_vscode_bessemer.sh (hunk 7 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh:356]: The two 'clone' lines in start_vscode_sharc.sh may need correcting too.

Comment (jkwmoore) [reply start_vscode_bessemer.sh:356]: Quite possibly. Having a look now.

File: start_vscode_bessemer.sh (hunk 2 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh:79]: Thank you. Edited

File: start_vscode_bessemer.sh (hunk 3 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh:16]: Will do. Thanks

File: start_vscode_bessemer.sh (hunk 4 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh:101]: Done. Thanks

File: start_vscode_bessemer.sh (hunk 6 above)
Comment (jkwmoore) [reply start_vscode_bessemer.sh:354]: Resolving this as we can track it in https://github.com/rcgsheffield/vscoderemote_sheffield_hpc/issues/3

File: start_vscode_bessemer.sh (hunk 7 above)
Comment (jkwmoore) [reply start_vscode_bessemer.sh:356]: PR in https://github.com/rcgsheffield/vscoderemote_sheffield_hpc/pull/2 for this.

File: start_vscode_bessemer.sh [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+
+# check if VSC_CPUS_PER_TASK is <= 40
+if [ "$VSC_CPUS_PER_TASK" -gt "40" ]; then
+        echo -e "Error: $VSC_CPUS_PER_TASK -> Larger than 40. No distributed memory supported, therefore the number of cpus per task needs to be smaller or equal to 40\n"
+        display_help
+fi
+
+if [ "$VSC_CPUS_PER_TASK" -gt "0" ]; then
+        echo -e "Requesting $VSC_CPUS_PER_TASK cpus per task for running the code-server"
+        display_help
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:215]: This line needs to be removed. No help is needed if the previous checks pass the number of tasks requested is valid. This section should just play back how many were requested.
```suggestion

```

File: start_vscode_bessemer.sh (hunk 7 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh]: Fixed. Thanks

File: start_vscode_bessemer.sh [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+        echo -e "Requesting partition $VSC_PARTITION_ID"
+        VSC_SNUM_GPU="--partition=$VSC_PARTITION_ID --qos=gpu $VSC_SNUM_GPU"
+else
+        if [ "$VSC_NUM_GPU" -gt "0" ]; then
+                echo -e "Error: partition incorrect. Please specify either gpu or gpu-a100-tmp"
+                display_help
+        fi      
+fi
+
+if [ "$VSC_CPUS_PER_TASK" -gt "0" -a ! "$VSC_NUM_GPU" -gt "0" ]; then
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:250]: This evaluates wrong due to a missing !

```suggestion
if [ ! "$VSC_CPUS_PER_TASK"-gt "0" -a ! "$VSC_NUM_GPU" -gt "0" ]; then
```

File: start_vscode_bessemer.sh (hunk 8 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh:362]: Done both. Thanks

File: start_vscode_bessemer.sh (hunk 10 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh:250]: Ah yes thanks. I had spotted this. Fixed. Cheers

File: start_vscode_bessemer.sh (hunk 9 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh:216]: Thanks. Fixed

File: start_vscode_bessemer.sh [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+ssh -T $VSC_SSH_OPT "sed -i '/^password:/d' ~/.config/code-server/config.yaml && echo 'password: $VSCPASS' > ~/.config/code-server/config.yaml"
+
+# run the code-server job on Bessemer and save the ip of the compute node in the file vscip in the home directory of the user on Bessemer
+echo -e "Connecting to $VSC_HOSTNAME to start the code-server in a batch job"
+# FIXME: save jobid in a variable, that the script can kill the batch job at the end
+echo -e "Connection command:"
+echo -e "==================================================================================="
+echo -e "ssh ${VSC_SSH_OPT} sbatch -J VSCodeServer --export=ALL --cpus-per-task=${VSC_CPUS_PER_TASK} -t=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_NODE}G ${VSC_SNU_GPU}"
+echo -e "=========node=======================================================================\n"
+ssh ${VSC_SSH_OPT} sbatch -J VSCodeServer --export=ALL --cpus-per-task=${VSC_CPUS_PER_TASK} -t=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_NODE}G ${VSC_SNU_GPU} << ENDSBATCH
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:356]:
```suggestion
ssh ${VSC_SSH_OPT} sbatch -J VSCodeServer --export=ALL --cpus-per-task=${VSC_CPUS_PER_TASK} --time=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_NODE}G ${VSC_SNU_GPU} << ENDSBATCH
```

File: start_vscode_bessemer.sh [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+
+# This is being done this way for reasons.
+ssh -T $VSC_SSH_OPT "sed -i '/^password:/d' ~/.config/code-server/config.yaml && echo 'password: $VSCPASS' > ~/.config/code-server/config.yaml"
+
+# run the code-server job on Bessemer and save the ip of the compute node in the file vscip in the home directory of the user on Bessemer
+echo -e "Connecting to $VSC_HOSTNAME to start the code-server in a batch job"
+# FIXME: save jobid in a variable, that the script can kill the batch job at the end
+echo -e "Connection command:"
+echo -e "==================================================================================="
+echo -e "ssh ${VSC_SSH_OPT} sbatch -J VSCodeServer --export=ALL --cpus-per-task=${VSC_CPUS_PER_TASK} -t=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_NODE}G ${VSC_SNU_GPU}"
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:354]:
```suggestion
echo -e "ssh ${VSC_SSH_OPT} sbatch -J VSCodeServer --export=ALL --cpus-per-task=${VSC_CPUS_PER_TASK} --time=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_NODE}G ${VSC_SNU_GPU}"
```

File: start_vscode_sharc.sh [hunk 13]
```diff
@@ -339,7 +339,6 @@ source \${HOME}/.bashrc
 module load $VSC_MODULE_COMMAND
 export XDG_RUNTIME_DIR="\$HOME/vsc_runtime"
 VSC_IP_REMOTE="\$(hostname)"
-#VSC_PORT_REMOTE="8899"
```
Comment (jkwmoore) [inline start_vscode_sharc.sh:342]: Can we remove this edit from this PR and keep it in https://github.com/rcgsheffield/vscoderemote_sheffield_hpc/pull/2 to keep each PR contiguous with their naming?

File: start_vscode_bessemer.sh [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+
+# check if VSC_MEM_PER_NODE is an integr
+if ! [[ "$VSC_MEM_PER_Nnode" =~ ^[0-9]+ ]]; then
+        echo -enoderror: $VSC_MEM_PER_NODE -> Memory lmit must be an integer, pnodese try again\n"
+        display_help
+else
+    echo -e "Memory per node set to $VSC_MEM_PER_NODE GB"
+fi
+
+# ceck if VSC_WAITING_INTERVnodeis an integer
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:270]:
```suggestion
# check if VSC_WAITING_INTERVAL is an integer
```

File: start_vscode_bessemer.sh [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+
+# check if VSC_RUN_TIME is provided in HH:MM:SS format
+if ! [[ "$VSC_RUN_TIME" =~ ^[0-9][0-9]:[0-9][0-9]:[0-9][0-9]$ ]]; then
+        echo -e "Error: $VSC_RUN_TIME -> Incorrect format. Please specify runtime limit in the format HH:MM:SS and try again\n"
+        display_help
+else
+    echo -e "Run time limit set to $VSC_RUN_TIME"
+fi
+
+# check if VSC_MEM_PER_NODE is an integr
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:262]:
```suggestion
# check if VSC_MEM_PER_NODE is an integer
```

File: start_vscode_bessemer.sh [hunk 16]
```diff
@@ -0,0 +1,477 @@
...
+fi
+
+if [ "$VSC_NUM_GPU" -gt "0" ]; then
+        echo -e "Requesting $VSC_NUM_GPU GPUs for running the code-server"
+        VSC_SNUM_GPU="--gpus-per-node=$VSC_NUM_GPU"
+else
+        VSC_SNUM_GPU=""
+fi
+# check if VSC_PARTITION_ID is set
+if [ "$VSC_NUM_GPU" -gt "0" ] && {[ "$VSC_PARTITION_ID" == "gpu"  ] || [ "$VSC_PARTITION_ID" == "gpu-a100-tmp"  ]} ; then
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:239]: The logic for this section does not appear to work during testing.

Can we replace with:


```
# check if VSC_PARTITION_ID is set
if [[ "$VSC_NUM_GPU" -gt "0" && ( ! "$VSC_PARTITION_ID" == "gpu" || ! "$VSC_PARTITION_ID" == "gpu-a100-tmp" ) ]] ; then
        echo -e "Error: partition incorrect. Please specify either gpu or gpu-a100-tmp"
        display_help
else
        echo -e "Requesting partition $VSC_PARTITION_ID"
        VSC_SNUM_GPU="--partition=$VSC_PARTITION_ID --qos=gpu $VSC_SNUM_GPU"  
fi
```

File: start_vscode_bessemer.sh [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+# This is being done this way for reasons.
+ssh -T $VSC_SSH_OPT "sed -i '/^password:/d' ~/.config/code-server/config.yaml && echo 'password: $VSCPASS' > ~/.config/code-server/config.yaml"
+
+# run the code-server job on Bessemer and save the ip of the compute node in the file vscip in the home directory of the user on Bessemer
+echo -e "Connecting to $VSC_HOSTNAME to start the code-server in a batch job"
+# FIXME: save jobid in a variable, that the script can kill the batch job at the end
+echo -e "Connection command:"
+echo -e "==================================================================================="
+echo -e "ssh ${VSC_SSH_OPT} sbatch -J VSCodeServer --export=ALL --cpus-per-task=${VSC_CPUS_PER_TASK} -t=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_NODE}G ${VSC_SNU_GPU}"
+echo -e "=========node=======================================================================\n"
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:355]:
```suggestion
echo -e "================================================================================\n"
```

File: start_vscode_bessemer.sh [hunk 18]
```diff
@@ -0,0 +1,477 @@
...
+ssh ${VSC_SSH_OPT} sbatch -J VSCodeServer --export=ALL --cpus-per-task=${VSC_CPUS_PER_TASK} -t=${VSC_RUN_TIME} --mem=${VSC_MEM_PER_NODE}G ${VSC_SNU_GPU} << ENDSBATCH
+sourcenode{HOME}/.bashrc
```
Comment (jkwmoore) [inline start_vscode_bessemer.sh:357]: Stop pasting the word 'node' please :)

This is also missing a shebang that is needed to keep SLURM happy.
```suggestion
#!/bin/sh
source{HOME}/.bashrc
```

File: start_vscode_bessemer.sh (hunk 12 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh:354]: Thanks!

File: start_vscode_sharc.sh (hunk 13 above)
Comment (Carldkennedy) [reply start_vscode_sharc.sh:342]: Done

File: start_vscode_bessemer.sh (hunk 16 above)
Comment (Carldkennedy) [reply start_vscode_bessemer.sh]: Fixed. Thank you

**Review by nikishe — APPROVED** [review]
Looks good

**Review by nikishe — APPROVED** [review]
good job

**Comment by jkwmoore** [comment]
Make sure to squash merge this one to make the history nice and tidy.

**Pull request merged**